### PR TITLE
Expand nested conditional operators back into if-else

### LINKED ISSUE #2027: [Enhancement] Optionally disable ternary operator for long statements
[sample.zip](https://github.com/icsharpcode/ILSpy/files/4707275/sample.zip) (part of [Terraria](https://store.steampowered.com/app/105600))
The code was decompiled "correctly" but as a 2000-characters long line:
```
((float)num > (float)(58 + player.armor.Length + player.dye.Length + player.miscEquips.Length + player.miscDyes.Length + player.bank.item.Length + player.bank2.item.Length + player.bank3.item.Length + 1)) ? player.bank4.item[num - 58 - (player.armor.Length + player.dye.Length + player.miscEquips.Length + player.miscDyes.Length + player.bank.item.Length + player.bank2.item.Length + player.bank3.item.Length + 1) - 1] : (((float)num > (float)(58 + player.armor.Length + player.dye.Length + player.miscEquips.Length + player.miscDyes.Length + player.bank.item.Length + player.bank2.item.Length + 1)) ? player.bank3.item[num - 58 - (player.armor.Length + player.dye.Length + player.miscEquips.Length + player.miscDyes.Length + player.bank.item.Length + player.bank2.item.Length + 1) - 1] : (((float)num > (float)(58 + player.armor.Length + player.dye.Length + player.miscEquips.Length + player.miscDyes.Length + player.bank.item.Length + player.bank2.item.Length)) ? player.trashItem : (((float)num > (float)(58 + player.armor.Length + player.dye.Length + player.miscEquips.Length + player.miscDyes.Length + player.bank.item.Length)) ? player.bank2.item[num - 58 - (player.armor.Length + player.dye.Length + player.miscEquips.Length + player.miscDyes.Length + player.bank.item.Length) - 1] : (((float)num > (float)(58 + player.armor.Length + player.dye.Length + player.miscEquips.Length + player.miscDyes.Length)) ? player.bank.item[num - 58 - (player.armor.Length + player.dye.Length + player.miscEquips.Length + player.miscDyes.Length) - 1] : (((float)num > (float)(58 + player.armor.Length + player.dye.Length + player.miscEquips.Length)) ? player.miscDyes[num - 58 - (player.armor.Length + player.dye.Length + player.miscEquips.Length) - 1] : (((float)num > (float)(58 + player.armor.Length + player.dye.Length)) ? player.miscEquips[num - 58 - (player.armor.Length + player.dye.Length) - 1] : (((float)num > (float)(58 + player.armor.Length)) ? player.dye[num - 58 - player.armor.Length - 1] : (((float)num > 58f) ? player.armor[num - 58 - 1] : player.inventory[num]))))))))
```
Adding an option to disable ternary operator `?:` would be great, either globally or only for long conditions.

### PR DESCRIPTION
**Pull request by siegfriedpammer**
`ExpressionTransforms.HandleConditionalOperator` collapses `if (c) a = x; else a = y;` into a
conditional operator, innermost first, and keeps going for as long as the chain does. A source
`else if` ladder therefore comes back as one expression, however long it was. The sample attached
to #2027 decompiles to a single **2095-character** statement; one NLog method reaches **2230
characters nested 29 brackets deep**.

`ExpandNestedConditionals` undoes that past one level, so a statement keeps at most a single `?:`.

```csharp
// before
text2 = ((!text.StartsWith("aspnet", ...) && !text.StartsWith("iis", ...)) ? (text.StartsWith("HostAppName", ...) ? ... /* 20 more levels */ ));

// after
if (text.StartsWith("aspnet", StringComparison.OrdinalIgnoreCase) || text.StartsWith("iis", StringComparison.OrdinalIgnoreCase))
{
    text2 += " - Extension NLog.Web.AspNetCore not included?";
}
else if (text.StartsWith("HostAppName", StringComparison.OrdinalIgnoreCase))
{
    ...
```

Note the compound assignment comes back too: the conditional form forced `text2 = ... (text2 + "...")`.

## Where it runs, and why that is the whole design

At the **end** of the pipeline, immediately before `AssignVariableNames` - not inside
`ExpressionTransforms`.

Every transform that needs its input to be a single expression has to see the collapsed form first:
object and collection initializers, `with`, switch expressions, interpolated string handlers, and
the query lambdas the C# stage later rewrites into clauses. Cutting the chain earlier leaves an
`if`/`else` between the statements they pattern-match on, and they silently stop matching - an
object initializer assigning an init-only member then does not even compile.

This is the same reasoning `ReduceNestingTransform` already gives for walking back
`ConditionDetection`'s aggressive else-inlining once the structure is settled: an earlier transform
is deliberately aggressive because later stages depend on it, and a late pass chooses the final
shape.

Two consequences of running late:

- **No type question.** Only a store to a declared local is expanded. Such a local carries its own
  type from metadata, so splitting the branches cannot widen the result to a stack type and lose a
  cast - the defect filed as #3736. No type inference is involved.
- **No polarity question.** The pass builds the `if`/`else` itself instead of inheriting
  `ConditionDetection`'s IL ordering, so the ladder reads `else if (cond)` rather than
  `else if (!(cond))`. Negated `else if` occurrences across the corpus are unchanged from master.

## Tests

**`TestCases/Pretty/ConditionalOperatorNesting.cs`** covers what must **not** change - each member
is a construct that has to stay a single expression, and each was a compile error or a silent
degradation in the earlier version of this PR:

| member | guards | guard |
|---|---|---|
| `ArrayInitializer` | array initializer block | |
| `Query` | query lambda | |
| `RefLocal` | ref local, CS8174 | `#if CS70` |
| `SwitchExpression` | switch expression arm | `#if CS80` |
| `ObjectInitializer` | init-only member, CS8852 | `#if CS90` |
| `WithExpression` | `with` / record clone, CS1061 | `#if CS90` |
| `FilterKeepsExpression` + the two ctor classes | catch-when filter, `: base(...)`, `: this(...)` | |
| `ShallowChainStaysInline`, `MismatchedBranchTypesStayInline`, `LogicOperatorsAreNotCounted` | the one-level boundary, and `&&` / `\|\|` sharing the IfInstruction representation | |

**`TestCases/ILPretty/ConditionalChain.il`** holds the positive case. It is an ILPretty test rather
than a Pretty one because a Pretty fixture is its own input and expected output, and a chain that
round-trips through collapse and expansion has no fixed point there - whichever polarity is written
comes back inverted.

## Measurements

decompdiff over 20 popular NuGet packages, all 20 decompilable on both sides, against
`origin/master`:

| | old | new |
|---|---|---|
| types compared | 5927 | |
| types changed | **35** | |
| new errors | **0** | |
| lines | 667345 | +451 |
| gotos / `//IL_` warnings / `<>` name leaks | | **0 delta** |

Across the changed types, deepest line 2230 -> 618 characters, maximum bracket depth 29 -> 7,
lines at depth >= 5 down from 22 to 12.

Every declaration the diff introduces carries a name from metadata - `XPathNavigator`, `DbType`,
`Expression`, `Type`, `string`, `byte[]`, `NamingStrategy` - except four (`int num`, `long num`)
that are pre-existing locals whose declaration split off its assignment. No temporary is typed
from a stack type.

`ICSharpCode.Decompiler.Tests`: **3587 total, 0 failed**.

## Scope

The transform handles a nested conditional in two situations.

**Stored to a variable that already exists** - the expansion reuses that variable, so nothing has
to be decided: the variable carries its own type from metadata.

**Anywhere else** - an argument, a return value, a field store - there is nothing to expand into,
so `ILExtraction` provides one, with `PrepareExtract` vetting the move so the order of evaluation
is preserved. The temporary it creates is typed from the *stack* type though, and `I4` is `int`,
`bool`, `char` and every enum at once; extracting on that basis turned a `bool` into `int num` with
`if (num == 0)`, and an enum into `dbType = (IsFixedLength ? 22 : 0)`.

`InferExpectedType` (new, next to `InferType` in `ILTypeExtensions`) is what makes that position
usable. Where `InferType` asks what a value *is*, it asks what the position the value flows into
says it *should be* - a parameter, a return type, a field - all of which carry their type in
metadata, and it answers `UnknownType` where nothing does. Extraction happens only where that
question has an answer, and the temporary is typed from it:

```csharp
XPathNavigator xPathNavigator;                  // the receiver's declaring type, no cast back
if (root) { xPathNavigator = node.GetRootElement(); } else if (IsElement) { … }
xPathNavigator.CreateAttribute("xmlns", prefix, …);

DbType dbType;                                  // the field's type, real member names
dbType = (IsFixedLength ? DbType.AnsiStringFixedLength : DbType.AnsiString);
```

Chains whose position names nothing are left alone - `JToken.ReadFrom`'s, for instance, sits inside
a `!(…)`, so the rule declines rather than guessing.

Two widenings were also built and rejected as unmeasurable: admitting stack slots whose type is not
`object`, and retyping a stack slot from the types its branches agree on. Both produced output
byte-identical to this PR over 5006 types.

Refs #2027 (implements the "only for long conditions" half of that request; there is no global
switch to disable `?:`).

### CODE REVIEW

File: ICSharpCode.Decompiler/IL/Transforms/ExpressionTransforms.cs [hunk 1]
```diff
@@ -602,6 +682,20 @@ IfInstruction HandleConditionalOperator(IfInstruction inst)
 			ILInstruction value1, value2;
 			if (trueInst.Instructions[0].MatchStLoc(out v, out value1) && falseInst.Instructions[0].MatchStLoc(v, out value2))
 			{
+				// An if-else-if chain collapses into one conditional operator per step, innermost first.
+				// Past a few levels the resulting `a ? x : (b ? y : (c ? z : w))` is harder to read than
+				// the if-else statements it came from, so stop collapsing and leave the rest as statements.
+				// Ordered cheapest and most selective first: only a chain that is already too deep
+				// is worth asking the more expensive questions about.
+				if (!inst.MatchLogicAnd(out _, out _) && !inst.MatchLogicOr(out _, out _)
```
Comment (christophwille) [inline ICSharpCode.Decompiler/IL/Transforms/ExpressionTransforms.cs]: **Bug (uncompilable output): object initializer with init-only property.** The retained if-chain sits between `stloc tmp(newobj)` and the setter call, so `TransformCollectionAndObjectInitializers` no longer matches and the setter is emitted as a plain assignment.

```csharp
public class Init { public int P { get; init; } }
public Init M(int n, int a, int b, int c) => new Init { P = n > 2 ? a : n > 1 ? b : c };
```
Expected (master):
```csharp
return new Init { P = ((n > 2) ? a : ((n > 1) ? b : c)) };
```
Actual (this PR):
```csharp
Init init = new Init();
int p;
if (n <= 2) { p = ((n > 1) ? b : c); } else { p = a; }
init.P = p;   // error CS8852: Init-only property 'Init.P' can only be assigned in an object initializer
return init;
```
Same shape with `required`:
```csharp
public class Req { public required int P { get; set; } }
public Req M(int n, int a, int b, int c) => new Req { P = n > 2 ? a : n > 1 ? b : c };
// actual: Req req = new Req(); int p; if (...) ...; req.P = p; return req;
// error CS9035: Required member 'Req.P' must be set in the object initializer
```

File: ICSharpCode.Decompiler/IL/Transforms/ExpressionTransforms.cs [hunk 2]
```diff
@@ -602,6 +682,20 @@ IfInstruction HandleConditionalOperator(IfInstruction inst)
...
+					&& (ReachesNestingDepth(value1, MaxConditionalOperatorNesting)
+						|| ReachesNestingDepth(value2, MaxConditionalOperatorNesting))
+					&& KeepsResultType(v, value1, context.TypeSystem)
+					&& KeepsResultType(v, value2, context.TypeSystem)
+					&& AllowsStatements(inst))
+				{
+					return inst;
```
Comment (christophwille) [inline ICSharpCode.Decompiler/IL/Transforms/ExpressionTransforms.cs]: **Bug (uncompilable output): `with` expression is lost entirely.** The clone call is emitted as the mangled `_003CClone_003E_0024()` and the init-only setter as a statement.

```csharp
public record Rec2 { public int P { get; init; } }
public Rec2 M(Rec2 i, int n, int a, int b, int c) => i with { P = n > 2 ? a : n > 1 ? b : c };
```
Expected:
```csharp
return i with { P = ((n > 2) ? a : ((n > 1) ? b : c)) };
```
Actual:
```csharp
Rec2 rec = i._003CClone_003E_0024();   // error CS1061
int p;
if (n <= 2) { p = ((n > 1) ? b : c); } else { p = a; }
rec.P = p;                             // error CS8852
return rec;
```

File: ICSharpCode.Decompiler/IL/Transforms/ExpressionTransforms.cs [hunk 3]
```diff
@@ -589,6 +589,86 @@ protected internal override void VisitIfInstruction(IfInstruction inst)
...
+
+		/// <summary>
+		/// Gets whether the target variable still describes both values once they are left in
+		/// separate statements. A stack slot has no type of its own and falls back to its stack
+		/// type, which loses the common type of the branches; the conditional operator infers it,
+		/// so in that case the collapsed form is the one that keeps the result well typed.
+		/// </summary>
+		static bool KeepsResultType(ILVariable v, ILInstruction value, ICompilation compilation)
+		{
+			if (v.Kind != VariableKind.StackSlot)
```
Comment (christophwille) [inline ICSharpCode.Decompiler/IL/Transforms/ExpressionTransforms.cs]: **Bug (uncompilable output): ref locals.** `KeepsResultType` returns `true` for every non-stack-slot variable, so a ref local assigned from a nested ref conditional is split into a declaration without initializer. The guard has no by-reference awareness.

```csharp
public int M(int n, int[] a, int[] b)
{
    ref int r = ref n > 2 ? ref a[0] : ref n > 1 ? ref b[0] : ref a[1];
    r++;
    return r;
}
```
Expected:
```csharp
ref int reference = ref n > 2 ? ref a[0] : ref n > 1 ? ref b[0] : ref a[1];
```
Actual:
```csharp
ref int reference;   // error CS8174: A declaration of a by-reference variable must have an initializer
if (n <= 2) { reference = ref n > 1 ? ref b[0] : ref a[1]; } else { reference = ref a[0]; }
```

File: ICSharpCode.Decompiler/IL/Transforms/ExpressionTransforms.cs [hunk 4]
```diff
@@ -589,6 +589,86 @@ protected internal override void VisitIfInstruction(IfInstruction inst)
...
+		/// lambdas of a query expression - and statements left behind there do not compile.
+		/// </summary>
+		bool AllowsStatements(IfInstruction inst)
+		{
+			// The innermost enclosing statement: an outer one would be the whole try instruction
+			// for a filter, whose block says nothing about the context the value is built in.
+			var statement = inst.Ancestors.FirstOrDefault(a => a.Parent is Block);
+			if (statement?.Parent is not Block block)
+				return false;
+			var options = ILInlining.OptionsForBlock(block, statement.ChildIndex, context);
```
Comment (christophwille) [inline ICSharpCode.Decompiler/IL/Transforms/ExpressionTransforms.cs]: **Bug: record struct primary ctor and static field initializers are no longer recognized.** `IsInConstructorInitializer` is false for struct ctors (no chained base call), so the if-chain stays in the ctor body and `RecordDecompiler` no longer detects the primary constructor.

```csharp
public record struct RS(int N, int A, int B, int C)
{
    public int X { get; init; } = N > 2 ? A : N > 1 ? B : C;
}
```
Expected: the positional record struct as written, with `X { get; init; } = (N > 2) ? A : ((N > 1) ? B : C);`.

Actual: a non-positional `record struct RS` with explicit `N`/`A`/`B`/`C` properties, an explicit `public RS(int N, int A, int B, int C)` ctor containing `int num; if (N <= 2) {...} X = num;`, and a `[CompilerGenerated] public readonly void Deconstruct(...)`.

Same mechanism with static fields (drops `beforefieldinit` semantics):
```csharp
public class St { static int a = 1, b = 2; public static int X = a > 2 ? a : b > 1 ? b : a; }
// actual: public static int X; static St() { int x; if (...) ...; X = x; }
```

File: ICSharpCode.Decompiler/IL/Transforms/ExpressionTransforms.cs [hunk 5]
```diff
@@ -589,6 +589,86 @@ protected internal override void VisitIfInstruction(IfInstruction inst)
...
+			return true;
+		}
+
+		/// <summary>
+		/// Gets whether the statement holding <paramref name="inst"/> may be split into several.
+		/// The pipeline forces aggressive inlining wherever the result has to stay a single
+		/// expression - constructor initializers, catch-when filters, expression trees and the
+		/// lambdas of a query expression - and statements left behind there do not compile.
+		/// </summary>
+		bool AllowsStatements(IfInstruction inst)
```
Comment (christophwille) [inline ICSharpCode.Decompiler/IL/Transforms/ExpressionTransforms.cs]: **Silent degradation in every other single-expression consumer.** None of these are `Aggressive` inlining contexts, so `AllowsStatements` says yes and the later transform no longer matches:

```csharp
// switch expression -> becomes a switch statement with `int result; if ...`
int S(int k, int n, int a, int b, int c) => k switch { 1 => n > 2 ? a : n > 1 ? b : c, 2 => a, _ => b };

// interpolated string -> explicit DefaultInterpolatedStringHandler h = new(5, 2); h.AppendLiteral("v="); int value; if ...; h.AppendFormatted(value); ...
string I(int n, int a, int b, int c) => $"v={(n > 2 ? a : n > 1 ? b : c)} w={n}";

// query -> xs.Where((int x) => { int num; if ...; return x > num; })
IEnumerable<int> Q(int[] xs, int n, int a, int b, int c) => from x in xs where x > (n > 2 ? a : n > 1 ? b : c) select x;

// array initializer -> int[] array = new int[2]; int num; if ...; array[0] = num; array[1] = 5;
int[] A(int n, int a, int b, int c) => new[] { n > 2 ? a : n > 1 ? b : c, 5 };
```
Call arguments, tuple literals and ctor args similarly get a temp. This is why the cut belongs after those transforms, not in `ExpressionTransforms`.

File: ICSharpCode.Decompiler/IL/Transforms/ExpressionTransforms.cs [hunk 6]
```diff
@@ -589,6 +589,86 @@ protected internal override void VisitIfInstruction(IfInstruction inst)
...
+		/// Gets whether the target variable still describes both values once they are left in
+		/// separate statements. A stack slot has no type of its own and falls back to its stack
+		/// type, which loses the common type of the branches; the conditional operator infers it,
+		/// so in that case the collapsed form is the one that keeps the result well typed.
+		/// </summary>
+		static bool KeepsResultType(ILVariable v, ILInstruction value, ICompilation compilation)
+		{
+			if (v.Kind != VariableKind.StackSlot)
+				return true;
+			if (value is not IfInstruction)
```
Comment (christophwille) [inline ICSharpCode.Decompiler/IL/Transforms/ExpressionTransforms.cs]: `KeepsResultType` is inconsistent rather than a type-safety check. `ILTypeExtensions.InferType` returns `UnknownType` for `ldc.i4`, `conv`, arithmetic and most instructions, and the O stack slot is `object`, so only chains whose leaves are plain int/long/float/double loads or calls are ever split:

```csharp
int x = n > 2 ? a[0] : n > 1 ? a[1] : a[2];          // split
int y = n > 2 ? 1 : n > 1 ? 2 : 3;                   // stays nested
int z = n > 2 ? a[0] + 1 : n > 1 ? a[1] + 1 : a[2] + 1; // stays nested
string s = n > 2 ? p : n > 1 ? q : r;                // stays nested (object slot vs string)
```
`MismatchedBranchTypesStayInline` passes because `ldstr` infers `Unknown`, not because of the `object` leaf.

File: ICSharpCode.Decompiler/IL/Transforms/ExpressionTransforms.cs [hunk 7]
```diff
@@ -602,6 +682,20 @@ IfInstruction HandleConditionalOperator(IfInstruction inst)
 			ILInstruction value1, value2;
 			if (trueInst.Instructions[0].MatchStLoc(out v, out value1) && falseInst.Instructions[0].MatchStLoc(v, out value2))
 			{
+				// An if-else-if chain collapses into one conditional operator per step, innermost first.
+				// Past a few levels the resulting `a ? x : (b ? y : (c ? z : w))` is harder to read than
+				// the if-else statements it came from, so stop collapsing and leave the rest as statements.
+				// Ordered cheapest and most selective first: only a chain that is already too deep
+				// is worth asking the more expensive questions about.
+				if (!inst.MatchLogicAnd(out _, out _) && !inst.MatchLogicOr(out _, out _)
+					&& (ReachesNestingDepth(value1, MaxConditionalOperatorNesting)
```
Comment (christophwille) [inline ICSharpCode.Decompiler/IL/Transforms/ExpressionTransforms.cs]: Dead code: `!inst.MatchLogicAnd(...) && !inst.MatchLogicOr(...)` on `inst` can never be false here, because both branches were just verified to be `Block`s while `MatchLogicAnd`/`MatchLogicOr` (PatternMatching.cs) require an `ldc.i4` branch. Likewise `depth <= 0` in `ReachesNestingDepth` is unreachable with the constant `1`, and `KeepsResultType`'s `value is not IfInstruction` early return duplicates the worklist's leaf case. With `MaxConditionalOperatorNesting = 1` the whole worklist reduces to `value is IfInstruction i && !i.MatchLogicAnd(..) && !i.MatchLogicOr(..)`. The comment sells the dead check as the cheap selective filter, which will mislead the next reader.

File: ICSharpCode.Decompiler.Tests/TestCases/Pretty/ConditionalOperatorNesting.cs [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+			{
+				num = b[0];
+			}
+			else if (n > 40)
+			{
+				num = a[1];
+			}
+			else if (n <= 30)
+			{
+				int num2;
```
Comment (christophwille) [inline ICSharpCode.Decompiler.Tests/TestCases/Pretty/ConditionalOperatorNesting.cs]: The fixture bakes in accidental shapes (this `num2` temporary, the stack-slot-only split, `roslynOnlyOptions` only) and has no negative case for any single-expression consumer: object initializer with init/required member, `with`, switch expression, ref local, query, interpolated string, positional record struct. Those are exactly where the change breaks (see the inline comments on `ExpressionTransforms.cs`); please add them here so the 5 compile errors above become red tests.

**Review by christophwille — DISMISSED** [review]
## Review summary

I built `ilspycmd` from this branch and from master, decompiled a small fixture assembly with both, and recompiled the output. Master's output compiles cleanly; this branch's output has **5 compiler errors** (CS8852 x2, CS1061, CS9035, CS8174) plus several silent degradations.

**Root cause:** the guard in `HandleConditionalOperator` cuts the chain in `ExpressionTransforms`, i.e. *before* every later transform that requires its input to be a single expression: object/collection initializers, `with`, switch expressions, query lambdas, interpolated-string handlers, record primary-constructor detection. `AllowsStatements` only consults `ILInlining.OptionsForBlock`, which knows about ctor initializers/filters/expression trees but nothing about those transforms. Any nested conditional feeding one of them now leaves an `if` chain between the statements they pattern-match on.

The 20-package corpus did not catch this because `KeepsResultType` only lets the guard fire when every leaf is a plain int/long/float/double load (`InferType` returns `Unknown` for `ldc.i4`, `conv`, arithmetic; an `object` stack slot never matches a `string` leaf). `MismatchedBranchTypesStayInline` passes for that reason, not for the one it claims.

**Suggested fix:** do not cut in `ExpressionTransforms` at all. Add a small late pass (after `TransformCollectionAndObjectInitializers` / switch-expression / query / interpolation transforms) that expands a *statement-level* `stloc local(if (c) x else y)` whose branches still nest a `?:` back into `if/else`, restricted to `VariableKind.Local` targets whose type is not by-reference. At that point every consumer that needed the expression has already taken it, the `else if` ladder this PR wants still appears, and the type / aggressive-inlining questions go away (a statement-level store to a plain local can always be split).

Repro samples with expected/actual are in the inline comments (all compiled Release, `net10.0`, decompiled with default settings). Please also add them as pretty-test negative cases so the guard cannot regress them again.

File: ICSharpCode.Decompiler.Tests/TestCases/PdbGen/NestedConditional.residual.txt [hunk 9]
```diff
@@ -0,0 +1,13 @@
+0x06000001 ICSharpCode.Decompiler.Tests.TestCases.PdbGen.NestedConditional.Ladder
```
Comment (dgrunwald) [inline ICSharpCode.Decompiler.Tests/TestCases/PdbGen/NestedConditional.residual.txt]: Why is this test useful?

*Reviewed with [Stampeded!](https://github.com/icsharpcode/Stampeded)*

File: ICSharpCode.Decompiler/IL/ILTypeExtensions.cs [hunk 10]
```diff
@@ -161,6 +163,46 @@ public static bool IsFloatType(this PrimitiveType type)
...
+		/// Where a value's own type is only its stack type - `I4` being `int`, `bool`, `char` and
+		/// every enum at once - the consumer often still knows, because a parameter, a return type
+		/// or a field carries its type in metadata.
+		/// </remarks>
+		public static IType InferExpectedType(this ILInstruction inst, int childIndex, ICompilation? compilation)
+		{
+			switch (inst)
+			{
+				case CallInstruction call:
+					bool hasThis = call.OpCode != OpCode.NewObj && !call.Method.IsStatic;
```
Comment (dgrunwald) [inline ICSharpCode.Decompiler/IL/ILTypeExtensions.cs]: Use `call.IsInstanceCall` instead.

**Review by dgrunwald — APPROVED** [review]
